tests/h/streamer/messages_test.py

```diff
@@ -363,6 +366,55 @@ def test_sends_if_in_group(
         )
         assert len(socket.send_json_payloads) == 1
 
+    @pytest.mark.parametrize(
+        "is_shared,socket_principals,expect_send",
+        [
+            # Only members of a private group get notified when shared annotations
+            # in that group are deleted.
+            (True, ["user:bob", "group:foobar"], True),
+            (True, ["group:foobar"], True),
+            (True, [security.Everyone], False),
+            # Only creators of private annotations get notified when they are
+            # deleted.
+            (False, ["user:bob", "group:foobar"], True),
+            (False, ["group:foobar"], False),
+            (False, [security.Everyone], False),
+        ],
+    )
+    def test_sends_annotation_delete_if_permissions_correct(
+        self,
+        fetch_annotation,
+        annotation_resource,
+        principals_allowed_by_permission,
+        is_shared,
+        socket_principals,
+        expect_send,
+    ):
+        message = {"annotation_id": "_", "action": "delete", "src_client_id": "pigeon"}
+        annotation = fetch_annotation.return_value
+        annotation.userid = "user:bob"
+        socket = FakeSocket("giraffe")
+        session = mock.sentinel.db_session
+        settings = {"foo": "bar"}
+
+        def principals(resource, permission):
+            if (
+                resource == annotation_resource.return_value.group
+                and permission == "read"
+            ):
+                return ["group:foobar"]
+            else:
+                return []
```

Comment:
There's a bit too much conditionality here to be clear about what's going on. I was expecting to see a `principals_allowed_by_permission.assert_called_once_with(...)` line here.

If that was combined with a canned response, I don't think we'd need this to be so active.